Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableEntitiesManager.cpp

```diff
@@ -316,161 +403,264 @@ namespace AzFramework
+
+                            spawnedEntities.push_back(
+                                CloneSingleEntity(*entitiesToSpawn[index], ticket.m_entityIdReferenceMap, *request.m_serializeContext));
+                            spawnedEntityIndices.push_back(index);
+                        }
+                    }
+                }
+                else
+                {
+                    for (uint32_t index : request.m_entityIndices)
+                    {
+                        if (index < entitiesToSpawn.size())
+                        {
+                            // If this entity has previously been spawned, give it a new id in the reference map
+                            RefreshEntityIdMapping(
+                                entitiesToSpawn[index].get()->GetId(), ticket.m_entityIdReferenceMap, ticket.m_previouslySpawned);
+
+                            auto aliasIt = AZStd::lower_bound(
+                                aliasBegin, aliasEnd, index,
+                                [](const Spawnable::EntityAlias& lhs, uint32_t rhs)
+                                {
+                                    return lhs.m_sourceIndex < rhs;
+                                });
+
+                            if (aliasIt == aliasEnd || aliasIt->m_sourceIndex != index)
+                            {
+                                spawnedEntities.emplace_back(
+                                    CloneSingleEntity(*entitiesToSpawn[index], ticket.m_entityIdReferenceMap, *request.m_serializeContext));
+                                spawnedEntityIndices.push_back(index);
+                            }
+                            else
+                            {
+                                // The list of entities has already been sorted and optimized (See SpawnableEntitiesAliasList:Optimize) so
+                                // can be safely executed in order without risking an invalid state.
+                                AZ::Entity* previousEntity = nullptr;
+                                do
+                                {
+                                    AZ::Entity* clone = CloneSingleAliasedEntity(
+                                        *entitiesToSpawn[index], *aliasIt, ticket.m_entityIdReferenceMap, previousEntity,
+                                        *request.m_serializeContext);
+                                    previousEntity = clone;
+                                    if (clone)
+                                    {
+                                        spawnedEntities.emplace_back(clone);
+                                        spawnedEntityIndices.push_back(index);
+                                    }
+
+                                    ++aliasIt;
+                                } while (aliasIt != aliasEnd && aliasIt->m_sourceIndex == index);
+                            }
```

Comment:
can you move this into a helper function because its the same code as in SpawnAllEntities() too.

Reply:
The shared code has already been factored into functions. The remaining code has some subtle differences because `SpawnAllEntities` can use a straight loop while `SpawnEntities` uses a random list of indices. I did try to split this up into further functions but it didn't help readability. To make it readable would come at the cost of performance which for entity spawning isn't worth taking DRY to an extreme imho.